### PR TITLE
Kahns's Algorithm in c++

### LINKED ISSUE #2271: Kahn's Algorithm
## 🚀 Feature

Kahn's algorithm is an algorithm for finding Topological sort for a graph. Since Topological sort is only possible for a Directed acyclic graph, this algorithm will also have a cycle detection logic. This is a BFS based approach.

### Have you read the [Contributing Guidelines on Pull Requests] 

Yes

## Motivation

I love data structures and algorithms. As a participant of gssoc 2021 I would like to contribute for the same .

## Pitch

The currently available algorithm for topological sort in C++ code base is a DFS based approach and will output some order even if the directed graph is cyclic. This algorithm checks for a cycle also and won't output any order instead will prompt that the graph has a cycle. 
Kindly assign me this under GSSOC 2021

### PR DESCRIPTION
**Pull request by dhruvinfo28**
### Have you read the [Contributing Guidelines on Pull Requests]
Yes

### Description

Added Kahn's algorithm.  closes #2271 

### Checklist

- [ Yes] I've read the contribution guidelines.
- [ Yes] I've checked the issue list before deciding what to submit.
- [ Yes] I've edited the `README.md` and link to my code.

### CODE REVIEW

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+3 4
+4 0
+Output:->
+No topological sort possible, there exists a cycle
+
+Time complexity: O(n+m) //Order of Sum of number of edges and vertices
+
+Auxillary Space: O(n)   //Order of number of vertices
+
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Add a new line at the EOF

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+3.
+Input:->
+5 5
+0 1
+1 2
+2 3
+3 4
+4 0
+Output:->
+No topological sort possible, there exists a cycle
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Only one sample I/O is required

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+int main(){
+    int n,m; cin>>n>>m;
+    vector <int> adj[n];
+    
+    int a,b;
+    for(int i=0;i<m;i++){
+        cin>>a>>b;
+        adj[a].push_back(b);
+    }
+    kahn_topological_sort(adj,n);
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Format the entire Code

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+4: Repeat 3 until the queue is empty.
+
+5: If count of visited nodes is not equal to the number of nodes in the graph then the topological sort is not possible for the given graph.
+*/
+
+#include<bits/stdc++.h>
+using namespace std;
+
+void kahn_topological_sort(vector <int> adj[], int n){
+    int in_degree[n]; //Array for keeping a track for indegree of all nodes
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Keep the Comments above the line of code not beside it

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    //Pushing nodes with in degree zero into the queue
+    for(int i=0;i<n;i++){
+        if(!in_degree[i]){
+            q.push(i);
+        }
+    }
+
+    int count_of_visited_nodes{};
+
+    vector <int> topological_order; //This vector will store the order
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Keep the Comments above the line of code not beside it

File: C-Plus-Plus/graphs/kahns_algorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        int curr = q.front();
+        topological_order.push_back(curr);
+        q.pop();
+        for(auto s: adj[curr]){
+            in_degree[s]--;
+            if(!in_degree[s]) q.push(s);
+        }
+        count_of_visited_nodes++;
+    }
+
```
Comment (HarshCasper) [inline C-Plus-Plus/graphs/kahns_algorithm.cpp]: Can you document it?

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/graphs/kahns_algorithm.cpp (hunk 3 above)
Comment (dhruvinfo28) [reply C-Plus-Plus/graphs/kahns_algorithm.cpp]: Do you mean removing extra lines? @HarshCasper

Comment (HarshCasper) [reply C-Plus-Plus/graphs/kahns_algorithm.cpp]: Yes. And to format the Code as well

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**